include/ignition/gazebo/components/ParticleEmitter.hh

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef IGNITION_GAZEBO_COMPONENTS_PARTICLEEMITTER_HH_
+#define IGNITION_GAZEBO_COMPONENTS_PARTICLEEMITTER_HH_
+
+#include <cstdint>
+#include <istream>
+#include <ostream>
+#include <string>
+
+#include <ignition/math/Color.hh>
+#include <ignition/math/Vector3.hh>
+#include <ignition/gazebo/components/Component.hh>
+#include <ignition/gazebo/components/Factory.hh>
+#include <ignition/gazebo/config.hh>
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace particles
+{
+  /// \brief All possible emitter type.
+  enum class EmitterType {POINT, BOX, CYLINDER, ELLIPSOID};
+
+  /// \brief Properties of a particle emitter.
+  /// A particle emitter also has:
+  ///    * a pose, which can be stored as a component of
+  ///      an emitter entity via ignition::gazebo::components::Pose
+  ///    * a name, which can be stored as a component of
+  ///      an emitter entity via ignition::gazebo::components::Name
+  ///    * a material, which can be stored as a component of
+  ///      an emitter entity via ignition::gazebo::components::Material
+  struct Emitter
```

Comment:
So far we haven't been creating entirely new classes to hold component data because there's usually already a class in some other library that can be used here and we don't duplicate information. I'd be curious to see the rest of the implementation to understand how this will fit with the rest of the code. Are we creating a message to hold this data? How about an SDF class? Another thing to keep in mind is that this header-only struct can't be extended with new fields without breaking ABI.

All this questioning is related to #494 :upside_down_face:

Reply:
@iche033 and I brainstormed about it when we were thinking about the particle integration and we thought this was the best alternative. I followed the same approach that we did with `LogicalAudio` in pull request #401 . 

The plan is to use an Ignition Message to store this data (there's already a [branch](https://github.com/ignitionrobotics/ign-msgs/blob/particle_emiter/proto/ignition/msgs/particle_emiter.proto) for it) and use a component that can be serialized to reach the clients. Then, in the rendering thread, use the Ignition Rendering API to create the particle emitter based on the request.

I wasn't planning to create an SDF class. Our idea is th

Reply:
We discussed whether or not we want to commit to adding this feature to the official sdf spec and create the DOM classes. But in the end, we thought adding to ign gazebo first seems like a less disruptive option that we could do in Dome for subt. 

I'm not opposed to adding to sdf spec. This could be a chance for something like `<visual><geometry><particles>` but we may want to include the sdf folks in the discussion / design

Reply:
>  I followed the same approach that we did with LogicalAudio

Oh ok I hadn't realized that was the approach there. My suggestion would be to use `ignition::msgs::ParticleEmitter` as the data for the component. That should be more efficient to serialize, and we don't need to worry about keeping the message and the component in sync. It looks like the new class isn't doing more than the message would, so I think there isn't a need for it.

> adding this feature to the official sdf spec

Yeah I think it makes sense to configure this feature through the `<plugin>` tag for now.

Reply:
Thanks for the suggestion, it looks better in the code to me. How about now? Is this what you were thinking?

Reply:
Much closer! Now that all the data is contained in the message, I think we could get rid of the `ignition::gazebo::particles::Emitter` class and just use the message directly in the component, right? Even if we think we may need to extend the class later, I think it would break ABI just like extending the message, so we can just extend the message.

When you use the message as the data, you don't need the custom operators, the custom serializer... It should be possible to define the component in one line like this:

https://github.com/ignitionrobotics/ign-gazebo/blob/a830b443989877b1e83170390f483b6c02952b25/include/ignition/gazebo/components/ExternalWorldWrenchCmd.hh#L41-L43

(also, let's not forget to bump the required `ign-msgs` version before merging this PR)

Reply:
Done in 4cbc9dd! How about now?